### ISSUE TITLE
Separable kernel based GP for Vector Valued output (output is defined over a grid)

### ISSUE THREAD

**Issue by PC-FSU** [cornellius-gp/gpytorch#2153]
Hi, I am trying to solve a problem where $x \in R^d$, and output $y$ is defined over a grid $s \in R^N$ (It's the output defined over a grid, not input). Now it may seem like a multi-output problem, but it's not a multi-output problem (As per I understand) since I am not modeling different components of output $y$ and calculating correlation among them; I am just modeling a vector output. Assuming that I have $n$ training points, $\{X,Y\} = [ (x_1,y_1),\cdots,\(x_n,y_n) ]$ , and each individual output is $N$ dimensional vector defined over a constant grid $s$, I could implement a naive GP, which would cost $\mathcal{O} (Nn)^3$. Now using separable kernel approach, I could represent the full kernel as $K = k_{x}(X) \bigotimes k_{s}(s)$, Where $k_{x},k_{s}$ are the kernels to capture co-var within input points and grid. This would reduce the computational cost to $\mathcal{O} (N^3 + n^3)$. `My question is: How to write a separable kernel-based GP using GpyTorch`. I have implemented a routine from scratch, which works fine, but doesn't scale well. `My next step is if I could implement a separable kernel using GpyTorch, I would like to extend this to a DKL-based model.` (I implement an MLP that models the mean prediction problem fine but provides no uncertainty estimate, that's why I want to explore DKL).

**Toy problem**:
 Function $f$ which takes a $x \in \mathcal{R}^d$ input, and return $y$ over a grid $s$ of size $N$.
$$f(x,s) = x_1 + sin(x_2*s)$$<br>

`X = [[0.1, 0.7],
       [0.3 , 0.4],
       [0.7, 0.8  ],
       [0.8, 0.8],
       [0.5, 0.1]]`
`n=5`
and
`s = np.linspace(0,2*np.pi,N), N = 63`.

The output would look like:
![image](https://user-images.githubusercontent.com/58872472/194180431-82717b3d-8ffa-4921-8fbb-4c8822866260.png)

  ```python
class Separable(gpytorch.models.ExactGP):
      def __init__(self, train_x, train_y, likelihood):
          super(Separable, self).__init__(train_x, train_y, likelihood)
          self.mean_module = gpytorch.means.ConstantMean()        
          self.data_covar_module = gpytorch.kernels.MaternKernel(nu=1.5, ard_num_dims=2)
          self.grid_covar_module = gpytorch.kernels.MaternKernel(nu=1.5, ard_num_dims=1)
        
      def forward(self, x, s):
              mean = self.mean_module(x)
              
              covar_x = to_linear_operator(self.data_covar_module(x))
              covar_s = to_linear_operator(self.grid_covar_module(s))
              covar = KroneckerProductLinearOperator(covar_x,covar_s)
              
              return gpytorch.distributions.MultivariateNormal(mean, covar)

likelihood = gpytorch.likelihoods.GaussianLikelihood()
 model = Separable((train_x,s), train_y, likelihood)



# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr = 0.1)  # Includes GaussianLikelihood parameters

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
training_iterations = 100

for i in range(training_iterations):
    optimizer.zero_grad()
    output = model(train_x,s)
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))
    optimizer.step()
```

Now I need to flatten my `train_y = Nn x 1` , since the kernel is of size $Nn \times Nn$, but when I call ` mean = self.mean_module(x)` I am working with only `x,` which has size `n.` I am confused who to proceed. Any help would be greatly appreciated

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [21], in <cell line: 12>()
     13 optimizer.zero_grad()
     14 output = model(train_x,s)
---> 15 loss = -mll(output, train_y)
     16 loss.backward()
     17 print('Iter %d/%d - Loss: %.3f' % (i + 1, training_iterations, loss.item()))

File ~\anaconda3\envs\GpyTorch\lib\site-packages\gpytorch\module.py:30, in Module.__call__(self, *inputs, **kwargs)
     29 def __call__(self, *inputs, **kwargs):
---> 30     outputs = self.forward(*inputs, **kwargs)
     31     if isinstance(outputs, list):
     32         return [_validate_module_outputs(output) for output in outputs]

File ~\anaconda3\envs\GpyTorch\lib\site-packages\gpytorch\mlls\exact_marginal_log_likelihood.py:63, in ExactMarginalLogLikelihood.forward(self, function_dist, target, *params)
     60     raise RuntimeError("ExactMarginalLogLikelihood can only operate on Gaussian random variables")
     62 # Get the log prob of the marginal distribution
---> 63 output = self.likelihood(function_dist, *params)
     64 res = output.log_prob(target)
     65 res = self._add_other_terms(res, params)

File ~\anaconda3\envs\GpyTorch\lib\site-packages\gpytorch\likelihoods\likelihood.py:65, in _Likelihood.__call__(self, input, *args, **kwargs)
     63 # Marginal
     64 elif isinstance(input, MultivariateNormal):
---> 65     return self.marginal(input, *args, **kwargs)
     66 # Error
     67 else:
     68     raise RuntimeError(
     69         "Likelihoods expects a MultivariateNormal input to make marginal predictions, or a "
     70         "torch.Tensor for conditional predictions. Got a {}".format(input.__class__.__name__)
     71     )

File ~\anaconda3\envs\GpyTorch\lib\site-packages\gpytorch\likelihoods\gaussian_likelihood.py:73, in _GaussianLikelihoodBase.marginal(self, function_dist, *params, **kwargs)
     71 noise_covar = self._shaped_noise_covar(mean.shape, *params, **kwargs)
     72 full_covar = covar + noise_covar
---> 73 return function_dist.__class__(mean, full_covar)

File ~\anaconda3\envs\GpyTorch\lib\site-packages\gpytorch\distributions\multivariate_normal.py:44, in MultivariateNormal.__init__(self, mean, covariance_matrix, validate_args)
     42 self.__unbroadcasted_scale_tril = None
     43 self._validate_args = validate_args
---> 44 batch_shape = torch.broadcast_shapes(self.loc.shape[:-1], covariance_matrix.shape[:-2])
     46 event_shape = self.loc.shape[-1:]
     48 # TODO: Integrate argument validation for LinearOperators into torch.distribution validation logic

File ~\anaconda3\envs\GpyTorch\lib\site-packages\linear_operator\operators\_linear_operator.py:2144, in LinearOperator.shape(self)
   2142 @property
   2143 def shape(self):
-> 2144     return self.size()

File ~\anaconda3\envs\GpyTorch\lib\site-packages\linear_operator\operators\_linear_operator.py:2137, in LinearOperator.size(self, dim)
   2131 def size(self, dim: Optional[int] = None) -> Union[torch.Size, int]:
   2132     """
   2133     Returns he size of the LinearOperator (or the specified dimension).
   2134 
   2135     :param dim: A specific dimension.
   2136     """
-> 2137     size = self._size()
   2138     if dim is not None:
   2139         return size[dim]

File ~\anaconda3\envs\GpyTorch\lib\site-packages\linear_operator\operators\sum_linear_operator.py:54, in SumLinearOperator._size(self)
     53 def _size(self):
---> 54     return torch.broadcast_shapes(*[lt.shape for lt in self.linear_ops])

File ~\anaconda3\envs\GpyTorch\lib\site-packages\torch\functional.py:125, in broadcast_shapes(*shapes)
    123             continue
    124         if result[i] != 1:
--> 125             raise RuntimeError("Shape mismatch: objects cannot be broadcast to a single shape")
    126         result[i] = shape[i]
    127 else:

RuntimeError: Shape mismatch: objects cannot be broadcast to a single shape
```

TLDR:
- `How to implement a separable kernel-based GP for a vector-valued output`
- `Can I use DKL for vector-valued output`

**Comment by gpleiss** [comment]
I fail to see how this is not a multi-output problem. The [multi-output GP model](https://docs.gpytorch.ai/en/stable/examples/03_Multitask_Exact_GPs/index.html#multi-output-vector-valued-functions) is specifically designed for modeling vector valued outputs, which is your use case as far as I am aware:

> Now it may seem like a multi-output problem, but it's not a multi-output problem (As per I understand) since I am not modeling different components of output y and calculating correlation among them; I am just modeling a vector output.

In the future, please also open up a [Discussion](https://github.com/cornellius-gp/gpytorch/discussions) for these types of questions: we are trying to reserve issues for bugs and feature requests.
